### ISSUE TITLE
pipline: binaries – should we build from source (in packager)?

### ISSUE THREAD

**Issue by graphicore** [googlefonts/gftools#231]
This came up when I recently used the packager to start the PR/QA processes for updates of some SMC fonts namely , [Manjari](https://github.com/google/fonts/pull/2572), [Gayathri](https://github.com/google/fonts/pull/2573) and [Chilanka](https://github.com/google/fonts/pull/2573). 

The main motivation was that SMC has no font binaries in their repos. However, they plan to attach font binaries to GitLab releases in the future, so this issue may not be terribly relevant. I'm posting this to collect thoughts and direction on how to go on.

There are two more issues that came up with this:

* accept GitLab/other git hosts #googlefonts/gf-docs/#75
* use GitHub (GitLab) release attachment binaries  # 230

For the three PRs referenced above I modified the `gftools packager` to accept a `build` field in it's upstream config, e.g:

https://github.com/google/fonts/blob/581699fcc0df7715a0f944e28b1a18d38e4aad62/ofl/chilanka/upstream.yaml#L23-L30

I think it's a nice workflow to have this in the packager. But: 

* This has security implications, e.g. in this case the `Makefile` in the repository could execute code that is malicious, e.g. when a build dependency on pypi.org was compromised (things like this happen). So, at least for a Font Bakery Dashboard application, just running any script is not good enough. And also anyone who executes the packager locally with an upstream configuration that has a `build` command should be aware that it may do unexpected things.
* The current approach is expecting a `bash` command, so perhaps limiting the usefulness e.g. on Windows, but IDK.

We could have a `gftools build` command, that would collect a couple of common build steps. That way we could be more secure and more portable and we could remove a lot of boilerplate from the build command above. A bit similar to what we see in the setup for GitHub Actions. Each possible step would be reviewed by us and could be used optionally (order could be configurable, having a dependency tree based implementation would be awesome). It would be cool if there would be an existing tool that we can use and just fill with content. Also, even `gftools build` tool would have limitations compared to just running any script, but that's the whole point.

Other suggestions are welcome, something simpler would be cool. ;-)

cc @santhoshtr, @davelab6, @m4rc1e, @chrissimpkins, @felipesanches

**Comment by chrissimpkins** [comment]
What is the final output from the packager? Archive of fonts and metadata file?

**Comment by graphicore** [comment]
> What is the final output from the packager? Archive of fonts and metadata file?

A PR to google/fonts.

**Comment by chrissimpkins** [comment]
Building production files from source makes alot of sense to me.  IMO this helps to enforce reproducible builds which is a desirable feature.  I really like your point about support for non GH remote repos too.  This relies on the availability of the full set of sources to get from source files to final production binaries.  Is that an issue across some/many/most projects?

For implementation security concerns, who will execute the packager? If it is the developer, they will be running their own source so this seems low risk.  If this is a G team reviewer, I imagine the workflow is that they have already reviewed the build and design source, then tested this locally on a blessed commit before it hits the packager? What scenario would involve someone who does not know / author the code executing it in order to submit a PR?

**Comment by graphicore** [comment]
> Building production files from source makes alot of sense to me. IMO this helps to enforce reproducible builds which is a desirable feature.

I agree, but is there a chain of arguments that establishes why reproducible builds are desirable. Right now, I think of:

 * automation, i.e. my main concern right now
 * the "chain of trust" argument, but we are far from establishing something like this
 * removal of black boxes, hence contributing to projects gets easier and the projects become de-facto FLOSS

> This relies on the availability of the full set of sources to get from source files to final production binaries. Is that an issue across some/many/most projects?

AFAIK, some projects in the past released only binaries not editor source files. We don't accept that any more. However, binaries built with proprietary compilers, e.g. GlyphsApp or FontLab, can be found more often. We are trying to get all projects to use `fontmake`, but I see also cases that are not that straight forward. My own Libre Barcode builds the UFO sources from Javascript-code before running fontmake. Many projects need to run commands after fontmake, like `gftools fix-*` scripts etc.


> For implementation security concerns, who will execute the packager? If it is the developer, they will be running their own source so this seems low risk.

This will likely not be a concern for us.

> If this is a G team reviewer, I imagine the workflow is that they have already reviewed the build and design source, then tested this locally on a blessed commit before it hits the packager?

I don't think it will be like this, especially for updates that already have an `upstream.yaml` it'll rather be: run the packager directly and see what QA brings up in the GitHub PR thread. Then maybe review the sources, if there are issues. We also shouldn't expect a review from a font engineer to discover potential attacks. And, even if a project builds locally, it doesn't mean nothing undesired was executed alongside.

> What scenario would involve someone who does not know / author the code executing it in order to submit a PR? 

The scenario above, but also potentially the Font Bakery Dashboard, which could be an interesting target for an attack, because it can potentially, at least, provide computing power (e.g. for DDOS or crypto-mining).

> The pypi dependency security risk applies to everyone with any build process and it doesn't appear that this approach increases that risk relative to not using a packaging tool that executes the build source.

OK, but the source repository itself is also a vector. A targeted attack could start with a) creating a libre font project, b) get it into google fonts, c) change the build script/build dependencies to do evil. Also, each maintainer of a source repository could become a target, via social engineering or an attacker could create a GitHub App that requires a GitHub Token with writing permissions for a libre font project repo, with that manipulate the source repository.

**Comment by graphicore** [comment]
#240 contains a provisional build feature which does not include any sandboxing other security features or any build in comfort. It has however disclaimers everywhere and must explicitly allowed using a flag (`-B/--allow-build`).

This is to unblock the SMC updates and I also believe having this provisionally is still better than having nothing i.e. a black box, where we don't know where the font binaries actually came from. google/fonts#2573,  google/fonts#2572 and already merged with `build` key: google/fonts#2574

**Comment by graphicore** [comment]
@simoncozens could you hava a look at this discussion and think about how to use `gftools builder` for this?

**Comment by davelab6** [comment]
I suspect we will need to support binary font releases, in zip packages,
from random URLs, in order to support automation of PRs from 100% of
upstreams.

**Comment by simoncozens** [comment]
gftools builder can already build e.g. Chilanka. (Manjari has a syntactically invalid `.designspace` file, oops.) I think the next step is to get it to run as a GitHub Action and either commit the binaries or upload them as artefacts.

**Comment by graphicore** [comment]
> I suspect we will need to support binary font releases, in zip packages, from random URLs, in order to support automation of PRs from 100% of upstreams.

This would the full black box again, that's why I rather strongly oppose it. You can put anything into those zip files, the sources, the commit history becomes irrelevant. Binaries in the repo at least imply some coherence with the commit history and the sources.

In order of preference:

1. build from sources
2. binaries in the repo
3. Nothing
4. binaries in zip files from random URLs

**Comment by graphicore** [comment]
Can I introduce the term "hot-fix disaster" for the time when we did hot-fixes on binaries and never fixed the sources/build scripts of these projects.

**Comment by simoncozens** [comment]
fontbakery literally has a check to make sure that the upstream is *not* a random binary zip file.

**Comment by vv-monsalve** [comment]
The Error `requests.exceptions.HTTPError: 401 Client Error: Unauthorized for url: https://api.github.com/graphql` is reported now when using the build option to update SMC's Chilanka font.

Is this option still available?

**Comment by eliheuer** [comment]
We figured it out, it was a PAT issue.

**Comment by vv-monsalve** [comment]
Thank you @eliheuer. Indeed it was a PAT issue
